src/futures/spi.rs

```diff
@@ -0,0 +1,295 @@
+//! Serial Peripheral Interface
+
+use core::future::Future;
+
+/// Full duplex (master mode)
+///
+/// # Notes
+///
+/// - It's the task of the user of this interface to manage the slave select lines
+///
+/// - Due to how full duplex SPI works each `read` call must be preceded by a `write` call.
+///
+/// - `read` calls only return the data received with the last `write` call.
+/// Previously received data is discarded
+///
+/// - Data is only guaranteed to be clocked out when the `read` call succeeds.
+/// The slave select line shouldn't be released before that.
+///
+/// - Some SPIs can work with 8-bit *and* 16-bit words. You can overload this trait with different
+/// `Word` types to allow operation in both modes.
+pub trait FullDuplex<Word> {
+    /// An enumeration of SPI errors
+    type Error;
+
+    /// The future associated with the `read` method.
+    type ReadFuture<'a>: Future<Output=Result<Word, Self::Error>> + 'a
+    where
+        Self: 'a;
+
+    /// The future associated with the `write` method.
+    type WriteFuture<'a>: Future<Output=Result<(), Self::Error>> + 'a
+    where
+        Self: 'a;
+
+    /// Reads the word stored in the shift register
+    ///
+    /// **NOTE** A word must be sent to the slave before attempting to call this
+    /// method.
+    fn read<'a>(&'a mut self) -> Self::ReadFuture<'a>;
+
+    /// Writes a word to the slave
+    fn write<'a>(&'a mut self, word: Word) -> Self::WriteFuture<'a>;
+}
+
+/// Clock polarity
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub enum Polarity {
```

Comment:
These should probably not be copypasted.

Reply:
It seems to me like there should be an agnostic place that has these kinds of items. Should the user just import them from blocking?

Reply:
given the blocking/nb/async split perhaps we could bubble them up to somewhere shared?